### PR TITLE
arch: move stack and task dump to common code

### PR DESCRIPTION
**Pull request by zyfeier**
## Summary

Move stack and task dump to common code

## Impact

Assert

## Testing

sabre-6quad:nsh

### CODE REVIEW

File: arch/misoc/src/lm32/lm32_registerdump.c [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: up_getusrsp
+ ****************************************************************************/
+
+#if CONFIG_ARCH_INTERRUPTSTACK > 3
+uintptr_t up_getusrsp(void)
+{
+  return g_intstacktop;
```
Comment (xiaoxiang781216) [inline arch/misoc/src/lm32/lm32_registerdump.c:51]: wrong

File: arch/misoc/src/lm32/lm32_registerdump.c [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: up_getusrsp
+ ****************************************************************************/
+
+#if CONFIG_ARCH_INTERRUPTSTACK > 3
```
Comment (xiaoxiang781216) [inline arch/misoc/src/lm32/lm32_registerdump.c:48]: remove the check

File: arch/mips/src/mips32/mips_registerdump.c [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: up_getusrsp
+ ****************************************************************************/
+
+#if CONFIG_ARCH_INTERRUPTSTACK > 3
+uintptr_t up_getusrsp(void)
+{
+  return g_intstacktop;
```
Comment (xiaoxiang781216) [inline arch/mips/src/mips32/mips_registerdump.c:52]: wrong

File: arch/mips/src/mips32/mips_registerdump.c [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: up_getusrsp
+ ****************************************************************************/
+
+#if CONFIG_ARCH_INTERRUPTSTACK > 3
```
Comment (xiaoxiang781216) [inline arch/mips/src/mips32/mips_registerdump.c:49]: remove the check

File: arch/mips/src/common/mips_getintstack.c [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+ ****************************************************************************/
+
+#if CONFIG_ARCH_INTERRUPTSTACK > 3
+
+uintptr_t up_get_intstackbase(void)
+{
+  return (uintptr_t)g_intstackalloc;
+}
+
+#endif
```
Comment (xiaoxiang781216) [inline arch/mips/src/common/mips_getintstack.c:46]: apply the similar change

File: arch/hc/src/m9s12/m9s12_registerdump.c [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: up_getusrsp
+ ****************************************************************************/
+
+#if CONFIG_ARCH_INTERRUPTSTACK > 3
+uintptr_t up_getusrsp(void)
+{
+  return g_intstacktop;
```
Comment (xiaoxiang781216) [inline arch/hc/src/m9s12/m9s12_registerdump.c:50]: wrong

File: arch/hc/src/m9s12/Make.defs [hunk 7]
```diff
@@ -25,10 +25,11 @@ CMN_CSRCS += hc_doirq.c hc_exit.c hc_idle.c hc_initialize.c
 CMN_CSRCS += hc_mdelay.c hc_modifyreg16.c hc_modifyreg32.c hc_modifyreg8.c
 CMN_CSRCS += hc_nputs.c hc_releasepending.c hc_releasestack.c
 CMN_CSRCS += hc_stackframe.c hc_udelay.c hc_switchcontext.c hc_usestack.c
+CMN_CSRCS += hc_getintstack.c
```
Comment (xiaoxiang781216) [inline arch/hc/src/m9s12/Make.defs:28]: merge to line 27

File: arch/hc/src/common/hc_getintstack.c [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+ ****************************************************************************/
+
+#if CONFIG_ARCH_INTERRUPTSTACK > 3
+
+uintptr_t up_get_intstackbase(void)
+{
+  return (uintptr_t)g_intstackalloc;
+}
+
+#endif
```
Comment (xiaoxiang781216) [inline arch/hc/src/common/hc_getintstack.c:46]: apply the similar change

File: arch/ceva/src/common/ceva_registerdump.c [hunk 9]
```diff
@@ -24,60 +24,42 @@
...
-  if (sp >= stack_base && sp < stack_base + stack_size)
-    {
-      stack = sp;
-    }
-  else
-    {
-      _alert("ERROR: Stack pointer is not within allocated stack\n");
-      stack = stack_base;
-    }
+  return CURRENT_REGS;
```
Comment (xiaoxiang781216) [inline arch/ceva/src/common/ceva_registerdump.c:45]: wrong

File: arch/ceva/src/common/ceva_registerdump.c [hunk 10]
```diff
@@ -24,60 +24,42 @@
...
  * Public Functions
  ****************************************************************************/
 
 /****************************************************************************
- * Name: z80_stackdump
+ * Name: up_getusrsp
  ****************************************************************************/
 
-void z80_stackdump(void)
+#if CONFIG_ARCH_INTERRUPTSTACK > 3
```
Comment (xiaoxiang781216) [inline arch/ceva/src/common/ceva_registerdump.c:42]: remove the check

File: arch/ceva/src/common/ceva_getintstack.c [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+ ****************************************************************************/
+
+#if CONFIG_ARCH_INTERRUPTSTACK > 3
+
+uintptr_t up_get_intstackbase(void)
+{
+  return (uintptr_t)g_intstackalloc;
+}
+
+#endif
```
Comment (xiaoxiang781216) [inline arch/ceva/src/common/ceva_getintstack.c:46]: apply the similar change

File: arch/avr/src/common/avr_getintstack.c [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+ ****************************************************************************/
+
+#if CONFIG_ARCH_INTERRUPTSTACK > 3
+
+uintptr_t up_get_intstackbase(void)
+{
+  return (uintptr_t)g_intstackalloc;
+}
+
+#endif
```
Comment (xiaoxiang781216) [inline arch/avr/src/common/avr_getintstack.c:46]: apply the similar change

File: arch/avr/src/avr32/avr_registerdump.c [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: up_getusrsp
+ ****************************************************************************/
+
+#if CONFIG_ARCH_INTERRUPTSTACK > 3
```
Comment (xiaoxiang781216) [inline arch/avr/src/avr32/avr_registerdump.c:45]: remove the check

File: arch/avr/src/avr/avr_registerdump.c [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: up_getusrsp
+ ****************************************************************************/
+
+#if CONFIG_ARCH_INTERRUPTSTACK > 3
```
Comment (xiaoxiang781216) [inline arch/avr/src/avr/avr_registerdump.c:45]: remove the check

File: arch/avr/src/avr/avr_checkstack.c [hunk 15]
```diff
@@ -164,6 +164,7 @@ ssize_t up_check_stack_remain(void)
 }
 
 #if CONFIG_ARCH_INTERRUPTSTACK > 3
+
```
Comment (xiaoxiang781216) [inline arch/avr/src/avr/avr_checkstack.c:167]: revert

File: arch/arm64/src/common/arm64_registerdump.c [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: up_getusrsp
+ ****************************************************************************/
+
+#if CONFIG_ARCH_INTERRUPTSTACK > 7
```
Comment (xiaoxiang781216) [inline arch/arm64/src/common/arm64_registerdump.c:50]: remove the check

File: arch/arm64/src/common/arm64_getintstack.c [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: up_get_intstackbase
+ ****************************************************************************/
+
+#if CONFIG_ARCH_INTERRUPTSTACK > 7
+
```
Comment (xiaoxiang781216) [inline arch/arm64/src/common/arm64_getintstack.c:40]: remove line 40 and 49

File: arch/arm64/src/common/arm64_checkstack.c [hunk 18]
```diff
@@ -217,15 +217,11 @@ ssize_t up_check_stack_remain(void)
 }
 
 #if CONFIG_ARCH_INTERRUPTSTACK > 7
+
```
Comment (xiaoxiang781216) [inline arch/arm64/src/common/arm64_checkstack.c:220]: revert the change

File: arch/arm/src/common/arm_registerdump.c [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: up_getusrsp
+ ****************************************************************************/
+
+#if CONFIG_ARCH_INTERRUPTSTACK > 3
```
Comment (xiaoxiang781216) [inline arch/arm/src/common/arm_registerdump.c:48]: remove the check

File: arch/arm/src/common/arm_getintstack.c [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: up_get_intstackbase
+ ****************************************************************************/
+
+#if CONFIG_ARCH_INTERRUPTSTACK > 3
+
```
Comment (xiaoxiang781216) [inline arch/arm/src/common/arm_getintstack.c:40]: remove line 40 and 49

File: arch/arm/src/common/arm_checkstack.c [hunk 21]
```diff
@@ -240,15 +240,11 @@ ssize_t up_check_stack_remain(void)
 }
 
 #if CONFIG_ARCH_INTERRUPTSTACK > 3
+
```
Comment (xiaoxiang781216) [inline arch/arm/src/common/arm_checkstack.c:243]: revert

File: sched/misc/assert.c [hunk 22]
```diff
@@ -121,6 +128,303 @@ static void assert_end(void)
...
+ * Name: up_stackdump
+ ****************************************************************************/
+
+static void up_stackdump(FAR uintptr_t sp, FAR uintptr_t stack_top)
+{
+  FAR uintptr_t stack;
+
+  for (stack = sp & ~0x1f; stack < (stack_top & ~0x1f); stack += 32)
+    {
+      FAR uint32_t *ptr = (uint32_t *)stack;
```
Comment (xiaoxiang781216) [inline sched/misc/assert.c:143]:
```suggestion
      FAR uint32_t *ptr = (FAR uint32_t *)stack;
```

File: sched/misc/assert.c [hunk 23]
```diff
@@ -121,6 +128,303 @@ static void assert_end(void)
     }
 }
 
+#ifdef CONFIG_ARCH_STACKDUMP
+
+/****************************************************************************
+ * Name: up_stackdump
+ ****************************************************************************/
+
+static void up_stackdump(FAR uintptr_t sp, FAR uintptr_t stack_top)
```
Comment (xiaoxiang781216) [inline sched/misc/assert.c:137]:
```suggestion
static void up_stackdump(uintptr_t sp, uintptr_t stack_top)
```

File: sched/misc/assert.c [hunk 24]
```diff
@@ -121,6 +128,303 @@ static void assert_end(void)
...
+{
+  FAR uintptr_t stack;
```
Comment (xiaoxiang781216) [inline sched/misc/assert.c:139]:
```suggestion
  uintptr_t stack;
```

File: sched/misc/assert.c [hunk 25]
```diff
@@ -121,6 +128,303 @@ static void assert_end(void)
...
+             ptr[4], ptr[5], ptr[6], ptr[7]);
+    }
+}
+
+/****************************************************************************
+ * Name: up_dump_stack
+ ****************************************************************************/
+
+static void up_dump_stack(FAR const char *tag, FAR uintptr_t sp,
+                          FAR uintptr_t base, uint32_t size,
```
Comment (xiaoxiang781216) [inline sched/misc/assert.c:157]:
```suggestion
                          uintptr_t base, uint32_t size,
```

File: sched/misc/assert.c [hunk 26]
```diff
@@ -121,6 +128,303 @@ static void assert_end(void)
...
+                          uint32_t used, bool force)
+{
+  FAR uintptr_t top = base + size;
```
Comment (xiaoxiang781216) [inline sched/misc/assert.c:160]:
```suggestion
  uintptr_t top = base + size;
```

File: sched/misc/assert.c [hunk 27]
```diff
@@ -121,6 +128,303 @@ static void assert_end(void)
...
+             stack, ptr[0], ptr[1], ptr[2], ptr[3],
+             ptr[4], ptr[5], ptr[6], ptr[7]);
+    }
+}
+
+/****************************************************************************
+ * Name: up_dump_stack
+ ****************************************************************************/
+
+static void up_dump_stack(FAR const char *tag, FAR uintptr_t sp,
```
Comment (xiaoxiang781216) [inline sched/misc/assert.c:156]:
```suggestion
static void up_dump_stack(FAR const char *tag, uintptr_t sp,
```

File: sched/misc/assert.c [hunk 28]
```diff
@@ -121,6 +128,303 @@ static void assert_end(void)
...
+                up_interrupt_context());
+  if (up_interrupt_context())
+    {
+      sp = up_getusrsp();
+    }
+#endif
+
+  up_dump_stack("User", sp,
+                (uintptr_t)rtcb->stack_base_ptr,
+                (uint32_t)rtcb->adj_stack_size,
```
Comment (xiaoxiang781216) [inline sched/misc/assert.c:225]: let's change all uint32_t to size_t for 64bit arch

File: sched/misc/assert.c [hunk 29]
```diff
@@ -121,6 +128,303 @@ static void assert_end(void)
...
+
+/****************************************************************************
+ * Name: up_dump_task
+ ****************************************************************************/
+
+static void up_dump_task(struct tcb_s *tcb, void *arg)
+{
+  char args[64] = "";
+#ifdef CONFIG_STACK_COLORATION
+  uint32_t stack_filled = 0;
```
Comment (xiaoxiang781216) [inline sched/misc/assert.c:261]: change all uint32_t to size_t

File: sched/misc/assert.c [hunk 30]
```diff
@@ -121,6 +128,303 @@ static void assert_end(void)
...
+         , args
+        );
+}
+
+/****************************************************************************
+ * Name: up_dump_backtrace
+ ****************************************************************************/
+
+#ifdef CONFIG_SCHED_BACKTRACE
+static void up_dump_backtrace(FAR struct tcb_s *tcb, FAR void *arg)
```
Comment (xiaoxiang781216) [inline sched/misc/assert.c:355]: let's remove up_ prefix from all function name

File: arch/x86_64/src/common/x86_64_getintstack.c [hunk 31]
```diff
@@ -0,0 +1,46 @@
...
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: up_get_intstackbase
+ ****************************************************************************/
+
+#if CONFIG_ARCH_INTERRUPTSTACK > 3
+
+uintptr_t up_get_intstackbase(void)
```
[the remark this answers is not in the document]
Comment (zyfeier) [reply arch/x86_64/src/common/x86_64_getintstack.c:40]: It's in intel64_regdump.c

File: sched/misc/assert.c [hunk 32]
```diff
@@ -121,6 +128,303 @@ static void assert_end(void)
...
+
+  if (cpuload.total > 0)
+    {
+      tmp      = (1000 * cpuload.active) / cpuload.total;
+      intpart  = tmp / 10;
+      fracpart = tmp - 10 * intpart;
+    }
+  else
+    {
+      intpart  = 0;
```
Comment (xiaoxiang781216) [inline sched/misc/assert.c:280]: merge to line 266-268

File: sched/misc/assert.c [hunk 33]
```diff
@@ -121,6 +128,303 @@ static void assert_end(void)
...
+#endif
+#ifdef CONFIG_SCHED_CPULOAD
+         "   %3zu.%01zu%%"
+#endif
+         "   %s%s\n"
+         , tcb->pid, tcb->sched_priority
+#ifdef CONFIG_SMP
+         , tcb->cpu
+#endif
+         , (unsigned long)tcb->adj_stack_size
```
Comment (xiaoxiang781216) [inline sched/misc/assert.c:332]: remove the cast, change lu to zu

File: sched/misc/assert.c [hunk 34]
```diff
@@ -121,6 +128,303 @@ static void assert_end(void)
...
+#ifdef CONFIG_STACK_COLORATION
+         , (unsigned long)up_check_tcbstack(tcb)
```
Comment (xiaoxiang781216) [inline sched/misc/assert.c:334]: remove the cast, change lu to zu

File: sched/misc/assert.c [hunk 35]
```diff
@@ -121,6 +128,303 @@ static void assert_end(void)
...
+        );
+#endif
+
+  nxsched_foreach(dump_task, NULL);
+
+#ifdef CONFIG_SCHED_BACKTRACE
+  nxsched_foreach(dump_backtrace, NULL);
+#endif
+}
+
```
Comment (xiaoxiang781216) [inline sched/misc/assert.c:427]: move assert_end here

File: arch/z80/src/ez80/Make.defs [hunk 36]
```diff
@@ -25,10 +25,6 @@ CMN_CSRCS += z80_idle.c z80_assert.c z80_doirq.c
 CMN_CSRCS += z80_mdelay.c z80_stackframe.c z80_udelay.c z80_usestack.c
 CMN_CSRCS += z80_nputs.c
 
-ifeq ($(CONFIG_ARCH_STACKDUMP),y)
-CMN_CSRCS += z80_stackdump.c
```
[the remark this answers is not in the document]
Comment (xiaoxiang781216) [reply arch/z80/src/ez80/Make.defs:29]: not remove z80_stackdump.c yet

**Review by xiaoxiang781216 — APPROVED** [review]

File: sched/misc/assert.c [hunk 37]
```diff
@@ -174,7 +473,20 @@ void _assert(FAR const char *filename, int linenum)
...
   syslog_flush();
 
-  up_assert(filename, linenum);
+  /* Show back trace */
+
+#ifdef CONFIG_SCHED_BACKTRACE
+  sched_dumpstack(running_task()->pid);
+#endif
+
+  up_assert();
```
Comment (CV-Bowen) [inline sched/misc/assert.c:482]: Seems wrong here, the dumped stack in showstacks() is different to the stack when register dumped in up_assert() due the function call and exit, which causes the stack changed. @xiaoxiang781216 @zyfeier

Comment (pkarashchenko) [reply sched/misc/assert.c:482]: What are the further actions here?

Comment (xiaoxiang781216) [reply sched/misc/assert.c:482]: @zyfeier is looking at this problem.

Comment (xiaoxiang781216) [reply sched/misc/assert.c:482]: > What are the further actions here?

@pkarashchenko here is the fix: https://github.com/apache/nuttx/pull/7955
